fix: replace broken centos 8 appstream repo with alma linux repos

supershal opened the pull request:
**What problem does this PR solve?**:


**Which issue(s) does this PR fix?**:
<!-- Add a link to the JIRA issue for both items below
* jql=key in (D2IQ-NUMBER)
-->
* https://jira.d2iq.com/browse/D2IQ-NUMBER


**Special notes for your reviewer**:
<!--
Use this to provide any additional information to the reviewers.
This may include:
- Manual testing steps.
- Best way to review the PR.
- Where the author wants the most review attention on.
- etc.
-->


**Does this PR introduce a user-facing change?**:
<!--
If yes, add a message in the 'release-note' block below.
If this PR fixes a COPS ticket, include it after the note like: "CLI: Some bug fix. (COPS-xxxx)"
-->
```release-note

```

dkoshkin commented on ansible/roles/packages/tasks/repo-redhat.yaml:
```diff
@@ -1,4 +1,49 @@
+    - name: Remove appstream repository
+      yum_repository:
+        name: appstream
+        state: absent
```
Maybe we need to update the cache?
```
fatal: [default]: FAILED! => {"changed": false, "msg": "Failed to download metadata for repo 'appstream': Cannot prepare internal mirrorlist: No URLs in mirrorlist", "rc": 1, "results": []}
```

https://teamcity.mesosphere.io/buildConfiguration/ClosedSource_KonvoyImageBuilder_E2eCentOS8/3341768?buildTab=log&focusLine=1336&linesState=469

supershal replied:
removed this bit. instead of adding it as "disabled" repo.

faiq commented on ansible/roles/packages/tasks/repo-redhat.yaml:
```diff
@@ -1,4 +1,49 @@
+      when:
+        - ansible_distribution_major_version == '8'
+
+    # The AppStream repo for Centos 8 is not available from centos mirror list.
+    # AlmaLinux is 1:1 binary compatible with RHEL and subscription free.
+    # Alma linux's extra repository symlinks epel repository rpms. https://wiki.almalinux.org/repos/AlmaLinux.html
+    # Why not RHEL repo?: RHEL is RHEL though the free subscription expires once a year.
+    # it cannot be renewed until it does expire and then you have to re-register your systems to pick up the new subscription.
+    # Also RHEL repos are missing some python3 rpms that we need.
```
great comment dude

dkoshkin commented:
Im a little confused by the epel-release changes, will this deprecation also impact konvoy1 https://github.com/mesosphere/konvoy/search?q=epel?

dkoshkin commented on ansible/roles/packages/tasks/repo-redhat.yaml:
```diff
@@ -1,4 +1,49 @@
+  when:
+    - not offline_mode_enabled
+  block:
+    - name: Remove appstream repository
```
```suggestion
    - name: remove appstream repository
```

supershal replied:
fixed it

dkoshkin commented on ansible/roles/packages/tasks/repo-redhat.yaml:
```diff
@@ -1,4 +1,49 @@
+    # Why not RHEL repo?: RHEL is RHEL though the free subscription expires once a year.
+    # it cannot be renewed until it does expire and then you have to re-register your systems to pick up the new subscription.
+    # Also RHEL repos are missing some python3 rpms that we need.
+    - name: Add Alma linux extra repository
```
```suggestion
    - name: add Alma linux extra repository
```

supershal replied:
fixed

supershal commented:
Wherever we use Centos 8 docker image or AMI, the "appstream" and "baseos" mirror list are broken. We have to either replace with some other repo or use fixed working "base_url" instead of mirror list.

faiq commented on ansible/roles/packages/tasks/repo-redhat.yaml:
```diff
@@ -1,4 +1,49 @@
+      when:
+        - ansible_distribution_major_version == '8'
+
+    # The correct epel-release rpms will be downloaded for Centos, RHEL and oracle 7 repos from in-built extra repos
+    # epel-release for Centos/RHEL/Oracle 8 will be installed using Alma linux "extra" repository
+    - name: install epel-release
+      yum:
+        name: epel-release
+        state: present
```
Where are the in-built extra repos coming from? Just a little confused Oracle7 seems to be failing on this for some reason

supershal replied:
in-built extra repos are installed with the OS in `/etc/yum.repos.d/` . each distribution has its own `base`, `extra` and `appstream` repos.  Those are not always in sync 100% with their `upstream` counter part.

supershal replied:
added back centos 7 repositories. The changes in this PR only affects centos 8 based repos.

supershal commented on ansible/roles/version/tasks/main.yaml:
```diff
@@ -16,11 +16,11 @@
     sed 's/$/{{ '_' + kubernetes_build_tag if kubernetes_build_tag }}/' |
-    sed 's/^\(.\+pause:.\+\)_{{ kubernetes_build_tag }}$/\1/'
+    sed 's/^\(.\+pause:.\+\)_{{ kubernetes_build_tag }}$/\1/' |
+    sed 's#^\(.\+coredns:\)\(.\+\)_{{ kubernetes_build_tag }}#{{ coredns_image_registry_repository }}:\2#'
```
a bit ugly fix to make `etcd` and `coredns` images work for both FIPS and non-FIPS versions. 
I intend to rewrite it in future.

supershal replied:
Non FIPS images:
```
(item=k8s.gcr.io/kube-apiserver:v1.21.6)
(item=k8s.gcr.io/kube-controller-manager:v1.21.6)
(item=k8s.gcr.io/kube-scheduler:v1.21.6)
(item=k8s.gcr.io/kube-proxy:v1.21.6)
(item=k8s.gcr.io/pause:3.4.1)
(item=k8s.gcr.io/etcd:3.4.13-0)
(item=k8s.gcr.io/coredns/coredns:v1.8.0)
(item=k8s.gcr.io/pause:3.2)
(item=docker.io/mesosphere/kfips:v0.1.0)
```

FIPS changes:
```
(item=docker.io/mesosphere/kube-apiserver:v1.21.6_fips.0)
(item=docker.io/mesosphere/kube-controller-manager:v1.21.6_fips.0)
(item=docker.io/mesosphere/kube-scheduler:v1.21.6_fips.0)
(item=docker.io/mesosphere/kube-proxy:v1.21.6_fips.0)
(item=docker.io/mesosphere/pause:3.4.1)
(item=docker.io/mesosphere/etcd:v3.4.13_fips.0)
(item=k8s.gcr.io/coredns/coredns:v1.8.0)
(item=k8s.gcr.io/pause:3.2)
(item=docker.io/mesosphere/kfips:v0.1.0)
```

supershal commented:
@faiq @dkoshkin The builds were breaking at multiple points because of two issues: Centos 8 deprecation and bug in pulling fips images. 
- replaces centos 8 repos with alma linux's extra repo
- added back logic to install epel repo for centos/oracle 7
- fixed the dynamic image parsing logic to pull correct version of etcd and coredns for both fips and non-fips mode
- removed makefile target for centos-8 and centos-8-nvidia buids

supershal commented:
Should we move "E2E CentOS 7 NVIDIA (Konvoy Image Builder" builds on Main branch only or nightly build? Or we can try to build images in different AWS region so that PRs not blocked by the lack of `p2` instances?

dkoshkin reviewed (approved):
Thanks!
